src/v/raft/consensus.cc

```diff
@@ -1200,215 +1200,190 @@ ss::future<> consensus::start() {
+        if (lstats.dirty_offset_term > _term) {
+            _term = lstats.dirty_offset_term;
+            _voted_for = {};
+        }
+        /**
+         * since we are starting, there were no new writes to the log
+         * before that point. It is safe to use dirty offset as a
+         * initial flushed offset since it is equal to last offset that
+         * exists on disk and was read in log recovery process.
+         */
+        _flushed_offset = lstats.dirty_offset;
+        /**
+         * The configuration manager state may be divereged from the log
+         * state, as log is flushed lazily, we have to make sure that
+         * the log and configuration manager has exactly the same
+         * offsets range
+         */
+        vlog(
+          _ctxlog.info, "Truncating configurations at {}", lstats.dirty_offset);
+
+        co_await _configuration_manager.truncate(
+          model::next_offset(lstats.dirty_offset));
+
+        /**
+         * We read some batches from the log and have to update the
+         * configuration manager.
+         */
+        if (st.config_batches_seen() > 0) {
+            co_await _configuration_manager.add(
+              std::move(st).release_configurations());
+        }
+
+        update_follower_stats(_configuration_manager.get_latest());
+
+        co_await _offset_translator.sync_with_log(_log, _as);
+
+        /**
+         * fix for incorrectly persisted configuration index. In
+         * previous version of redpanda due to the issue with
+         * incorrectly assigned raft configuration indicies
+         * (https://github.com/redpanda-data/redpanda/issues/2326) there
+         * may be a persistent corruption in offset translation caused
+         * by incorrectly persited configuration index. It may cause log
+         * offset to be negative. Here we check if this problem exists
+         * and if so apply necessary offset translation.
+         */
+        const auto so = start_offset();
+        const auto delta = _configuration_manager.offset_delta(start_offset());
+        // no prefix truncation was applied we do not need adjustment
+        if (so > model::offset(0) || so < delta) {
+            // if start offset is smaller than offset delta we need to apply
+            // adjustment
+            const configuration_manager::configuration_idx new_idx(so());
+            vlog(
+              _ctxlog.info,
+              "adjusting configuration index, current start offset: {}, "
+              "delta: {}, new initial index: {}",
+              so,
+              delta,
+              new_idx);
+            co_await _configuration_manager.adjust_configuration_idx(new_idx);
+        }
+
+        auto next_election = clock_type::now();
+        // set last heartbeat timestamp to prevent skipping first
+        // election
+        _hbeat = clock_type::time_point::min();
+        auto conf = _configuration_manager.get_latest().current_config();
```

Comment:
not sure if this is significant but in the continuation-based version this code is different

Reply:
the change is related with the fact that we no longer have `brokers()` in new configuration (v23.2.x)